src/unity/toolkits/tcmps/cnn_module.hpp

```diff
@@ -0,0 +1,61 @@
+/* Copyright © 2018 Apple Inc. All rights reserved.
+ *
+ * Use of this source code is governed by a BSD-3-clause license that can
+ * be found in the LICENSE.txt file or at https://opensource.org/licenses/BSD-3-Clause
+ */
+
+#ifndef TURI_MPS_CNN_MODULE_HPP_
+#define TURI_MPS_CNN_MODULE_HPP_
+
+#include <memory>
+
+#include "mps_float_array.hpp"
+
+namespace turi {
+namespace mps {
+
+/**
+ * A pure virtual interface for neural networks, used to abstract across model
+ * architectures and backend implementations.
+ */
+class cnn_module {
+public:
+  virtual ~cnn_module() = default;
+
+  /**
+   * Creates an object detection network using a backend appropriate to the
+   * current platform and hardware.
+   *
+   * \todo Define a object_detector_config struct to encapsulate these
+   *       parameters in a more self-documenting and typesafe way.
+   */
+  static std::unique_ptr<cnn_module> create_object_detector(
+      int n, int c_in, int h_in, int w_in, int c_out, int h_out, int w_out,
+      const float_array_map& config, const float_array_map& weights);
+
+  /**
+   * Sets the learning rate to be used for future calls to train.
+   */
+  virtual void set_learning_rate(float lr) = 0;
+
+  /**
+   * Performs one forward-backward pass.
+   */
+  virtual deferred_float_array train(const float_array& input_batch,
+                                     const float_array& label_batch) = 0;
+
+  /**
+   * Performs a forward pass.
+   */
+  virtual deferred_float_array predict(const float_array& input_batch) = 0;
```

Comment:
Can this be a const method too?

Reply:
I really don't know what would be best here.

Could we and would it make sense to mark the member variables which get mutated as `mutable`?

Reply:
I went ahead and made this method const, since I personally like the operational definition of "does not change the behavior of future method calls".

In this case, there's no need to mark any data structures as mutable, given C++ semantics. (I don't like relying 100% on the compiler to define const, since the following compiles:
```
class foo {
public:
  int read() const { return *x_; } 
  void increment() const {  // O RLY?
    *x_ += 1;
  }
private:
  int* x = new int(0);
};
```
)